dotnet/src/Skills/Skills.Web/Bing/BingConnector.cs

```diff
@@ -75,9 +80,29 @@ public async Task<IEnumerable<string>> SearchAsync(string query, int count = 1,
 
         BingSearchResponse? data = JsonSerializer.Deserialize<BingSearchResponse>(json);
 
-        WebPage[]? results = data?.WebPages?.Value;
-
-        return results == null ? Enumerable.Empty<string>() : results.Select(x => x.Snippet);
+        List<T>? returnValues = new();
+        if (data?.WebPages?.Value != null)
+        {
+            if (typeof(T) == typeof(string))
+            {
+                WebPage[]? results = data?.WebPages?.Value;
+                returnValues = results?.Select(x => x.Snippet).ToList() as List<T>;
+            }
+            else if (typeof(T) == typeof(WebPage))
+            {
+                List<WebPage>? webPages = new();
+                foreach (var webPage in data?.WebPages?.Value)
+                {
+                    webPages.Add(webPage);
+                }
+                returnValues = webPages.Take(count).ToList() as List<T>;
+            }
+            else
+            {
+                throw new NotSupportedException($"Type {typeof(T)} is not supported.");
+            }
```

Comment:
Agreed, this should be considered, however, it will be some time before the suggested change to allow complex types is implemented.

I feel that there is a benefit for some to be able to use this PR for now, considering returning just a "snippet" can only really be used as context for chained input to the LLM. Whereas some would prefer to simply use the skill to return results of a search to the user. This PR caters for both scenarios.

I'll address the warnings :)

Reply:
> considering returning just a "snippet" can only really be used as context for chained input to the LLM. Whereas some would prefer to simply use the skill to return results of a search to the user.

Well, I believe this is the purpose of these connectors in Semantic Kernel SDK - to be used as context for chained input to the LLM :)
If there is a need to get more complex responses from Bing/Google, it's possible to use Bing/Google API/SDK directly. 

I agree that complex responses have benefits, and we will definitely add them, as soon as Kernel will know how to work with them properly.

Reply:
> If a Skill exists in a library that is already in use, surely re-use of that library is more efficient

It means that we will need to implement Bing API completely, support and maintain it. But we don't want to re-invent the wheel, especially if such library already exists:
https://learn.microsoft.com/en-us/bing/search-apis/bing-web-search/quickstarts/sdk/web-search-client-library-csharp

Maybe it would be better to re-use already existing SDK in this connector, but on the other hand, it will bring more dependencies and functionality, that is not going to be used by Semantic Kernel (yet). Which is why this connector is lightweight at the moment, and we will extend it as soon as Kernel will know how to work with complex types.

Reply:
Forgive me, but I am a little confused here.

I have not changed the implementation of the API request in this Skill / connector at all (which equates to a simple Rest API call to the Bing Search API endpoint and is as lightweight as it gets). 

I have simply added an (optional) alternative output from the connector back to the caller, the original functionality should not be impacted at all.

I fail to see where the issue is here, beyond the response not being in a compatible format to use with the Kernel in a chaining scenario?

As can be seen from the example taken from our prod code (Teams bot)... The kernel works as expected.

```public async Task<BotResponse> Search(string text)
        {
            // Init the bingConnector and import the skill
            var bingConnector = new BingConnector(_config["BingSearchAPIKey"], uri: _config["BingSearchUrl"]);
            var searchSkill = _sk.ImportSkill(new WebSearchEngineSkill(bingConnector), "bing");

            ContextVariables contextVars = new($"{text}");
            contextVars.Set(WebSearchEngineSkill.Parameters.CountParam, $"3");
            contextVars.Set(WebSearchEngineSkill.Parameters.OffsetParam, $"0");

            // Perform the search and return the snippet(s) only.
            // var search = await _sk.RunAsync(contextVars, searchSkill["Search"]);
            // var snippets = JsonConvert.DeserializeObject<List<string>>(search.Result);

            // Perform the search and return the full result.
            var search = await _sk.RunAsync(contextVars, searchSkill["GetSearchResults"]);
            var WebPages = JsonConvert.DeserializeObject<List<WebPage>>(search.Result);

            string results = $"Bing Search Results\n\n" +
                    "----\n\n";

            switch(WebPages.Count)
            {
                case 0:
                    results += "No results found.";
                    break;
                    case 1:
                    results += $"Page Name: {WebPages[0].Name}\n\n" +
                        $"Page Snippet: {WebPages[0].Snippet}...\n\n" +
                        $"Page Url: {WebPages[0].Url}";
                    break;
                default:
                    for (int i = 0; i < WebPages.Count; i++)
                    {
                        results += $"--- Result {i+1} ---\n\n" +
                            $"Page Name: {WebPages[i].Name}\n\n" +
                            $"Page Snippet: {WebPages[i].Snippet}...\n\n" +
                            $"Page Url: {WebPages[i].Url}";

                        if (i+1 < WebPages.Count)
                        {
                            results += $"\n\n";
                        }
                    }
                    break;
            }
```
As can be seen, this example matches the format as defined in existing Skill examples found in the repo IE: [graph-api-skills](https://github.com/microsoft/semantic-kernel/blob/main/samples/dotnet/graph-api-skills/Program.cs) & is also compatible with the example provided for the Bing Search Skill itself: [10-BingSearch-using-kernel](https://github.com/microsoft/semantic-kernel/blob/main/samples/notebooks/dotnet/10-BingSearch-using-kernel.ipynb)

Reply:
> I have simply added an (optional) alternative output from the connector back to the caller, the original functionality should not be impacted at all.
> 
> I fail to see where the issue is here, beyond the response not being in a compatible format to use with the Kernel in a chaining scenario?

@SOE-YoungS An (optional) alternative output it's always good, but if it's not going to be used in the context of Semantic Kernel, then in my opinion it should live in another place. Skills functionality presented in this repository should be pluggable to Semantic Kernel.

But that's only my opinion :) If you disagree with it - that's totally fine.

A couple of suggestions from me based on recent changes:
1. Instead of `Newtonsoft.Json` you can use `System.Text.Json` if needed, to avoid extra dependency.
2. If you want to add an alternative optional output to Bing/Google connectors, you can add new separate methods in these classes without modifying already existing methods and `IWebSearchEngineConnector`.
3. It looks like you suppressed all the warnings, but we should avoid suppression and design a code in a way that it won't generate any warnings.

Thank you for contribution and sorry if my comments confused you.